Integrate DHCP/DNS into Smart Proxies

maximiliankolb opened the pull request:
* [x] I am familiar with the [contributing](https://github.com/theforeman/foreman-documentation/blob/master/CONTRIBUTING.md) guidelines.

Please cherry-pick my commits into:

* [x] Foreman 3.3/Katello 4.5
* [x] Foreman 3.2/Katello 4.4
* [x] Foreman 3.1/Katello 4.3

maximiliankolb commented:
See also https://github.com/theforeman/theforeman.org/pull/1916 & https://github.com/theforeman/foreman-documentation/pull/977

ekohl commented on guides/common/modules/proc_configuring-dns-on-smart-proxies.adoc:
```diff
@@ -0,0 +1,257 @@
+Extra providers are available as plugins and can be installed through packages. See the following pages for more information:
+
+* http://projects.theforeman.org/projects/foreman/wiki/List_of_Smart-Proxy_Plugins[List of smart proxy plugins]
+* link:%7B%7B%20site.baseurl%20%7D%7Dplugins/[Plugin installation documentation]
```
This looks like something went wrong in the conversion.

maximiliankolb replied:
Fixed.

ekohl commented on guides/common/modules/proc_configuring-dns-on-smart-proxies.adoc:
```diff
@@ -0,0 +1,257 @@
+
+Extra providers are available as plugins and can be installed through packages. See the following pages for more information:
+
+* http://projects.theforeman.org/projects/foreman/wiki/List_of_Smart-Proxy_Plugins[List of smart proxy plugins]
```
```suggestion
* https://projects.theforeman.org/projects/foreman/wiki/List_of_Smart-Proxy_Plugins[List of smart proxy plugins]
```

ekohl reviewed (commented):
Thanks for picking this up again. I do wonder if we should continue to document the manual steps. I like having it documented, but I feel like it should be a reference guide instead. Most users should use the installer instead. AFAIK we already have some guides for that.

ekohl commented on guides/common/assembly_performing-additional-configuration-on-capsule-server.adoc:
```diff
@@ -27,5 +27,9 @@ include::modules/proc_enabling-power-management-on-managed-hosts.adoc[leveloffse
 // Configuring DNS, DHCP, and TFTP on {SmartProxyServer}
 include::modules/proc_configuring-dns-dhcp-and-tftp.adoc[leveloffset=+1]
 
+include:modules/proc_configuring-dhcp-on-smart-proxies.adoc[leveloffset=+2]
+
+include:modules/proc_configuring-dns-on-smart-proxies.adoc[leveloffset=+2]
```
I still think these are reference guides, not to be consumed by most users since they don't use the installer. We have `guides/common/assembly_using-infoblox-as-dhcp-and-dns-providers.adoc` and `guides/common/modules/proc_configuring-dns-dhcp-and-tftp.adoc` as more generic instructions. Ideally we expand those with the additional providers.

If you want, I can work with you to make that happen.

maximiliankolb replied:
I've split the procedures.

mjivraja commented on guides/common/modules/proc_configuring-dhcp-on-smart-proxies.adoc:
```diff
@@ -0,0 +1,248 @@
+This requires it to be run either on the same host as the DHCP server or to have network filesystem access to these files.
+
+.Procedure
+. Set the path fo the _config_ and _leases_ files in the `dhcp_isc.yml` configuration file.
```
Possible typo 'fo'

maximiliankolb replied:
Fixed!

mjivraja commented on guides/common/modules/proc_configuring-dhcp-on-smart-proxies.adoc:
```diff
@@ -0,0 +1,248 @@
+:disable_ddns: true
+----
++
+By default, `disable_ddns` is set to true and dynamic DNS updates are disabled for all hosts that the {SmartProxy} creates.
```
```suggestion
For all the hosts that {SmartProxy} creates, “disable_ddns” is set to true, and dynamic DNS updates are disabled by default.
```
Trying to make it more simple.

maximiliankolb replied:
Slightly adjusted and applied.

mjivraja commented on guides/common/modules/proc_configuring-dns-on-smart-proxies.adoc:
```diff
@@ -0,0 +1,267 @@
+Your {SmartProxy} can manage DNS within its network.
+DNS management contains updating and removing DNS records from existing DNS zones.
+
+After your have enabled DNS, your {SmartProxy} can manipulate any DNS server that complies with the ISC Dynamic DNS Update standard.
```
```suggestion
After you have enabled DNS, your {SmartProxy} can manipulate any DNS server that complies with the ISC Dynamic DNS Update standard.
```
Possible typo.

maximiliankolb replied:
Fixed!

maximiliankolb reviewed (commented):
Applied your feedback and rebased to master.

mjivraja commented:
> Applied your feedback and rebased to master.

Thank you and done from my side. :+1:

maximiliankolb commented:
@ekohl Do you have concerns about the content itself or the structure? Before merging, I definitely need to split up the files into "one precedure per file"; then we're much more flexible on how/where to include them.

ekohl reviewed (changes requested):
I wanted to read through the preview, but it doesn't render.

As for the content. I still think this is not the right way. We should not be documenting the direct modification of config files in our primary guides. Any time a primary guide tells users to write a YAML file, it's wrong IMHO.

For example, `dhcp_isc.yml` is managed by the installer. Any local changes will be overwritten. https://docs.theforeman.org/nightly/Installing_Proxy/index-katello.html#configuring-dns-dhcp-and-tftp_smart-proxy already covers the default case (which is `dhcp_isc`). For background: the installer has `--foreman-proxy-dhcp-provider`, which defaults to `isc` and results in `:use_provider: dhcp_isc` in `dhcp.yml`.

However, you can use `--foreman-proxy-dhcp-provider native_ms`. Specifically, `native_ms` is not supported by the installer so you would indeed need to manage `dhcp_native_ms.yml` yourself. In my view this means that the `native_ms` provider is unsupported from a documentation perspective. If it would be supported, the installer has options to manage it. Note that there is `--foreman-proxy-dhcp-server` to manage the `server` value in `dhcp.yml`.

Then for `dhcp_libvirt` it would be `--foreman-proxy-dhcp-provider libvirt`. To configure network and url you use `--foreman-proxy-libvirt-network` and `--foreman-proxy-libvirt-connection`.

For DNS it's pretty much the same. There is `--foreman-proxy-dns-provider`, defaulting to `nsupdate`. You can probably predict that it results in `:use_provider: dns_nsupdate` in `dns.yml`.

You may wonder how to find this out. I usually find it easier to look in the Puppet module we use. We have [templates](https://github.com/theforeman/puppet-foreman_proxy/tree/master/templates) and follow the convention where the filenames match. A crude way to translate things is that `foreman_proxy::dns_tsig_principal` translates to the variable `dns_tsig_principal` on the `foreman_proxy` class. The installer exposes class parameters (which are variables but a variable doesn't have to a parameter). So you can read [manifests/init.pp](https://github.com/theforeman/puppet-foreman_proxy/blob/master/manifests/init.pp). We don't generate a reference (for technical reasons), but in this case you can also read `man foreman-installer` (which is the man page for the `foreman` scenario, faster than `foreman-installer --help`.

I hope this helps a bit. It's not complete, but hopefully it helps.

As for document structure: I don't like that https://docs.theforeman.org/nightly/Installing_Proxy/index-katello.html#configuring-dns-dhcp-and-tftp_smart-proxy is one huge section. Personally I think it would be better to document:
* Configuring DHCP
  * dhcp_isc
  * dhcp_libvirt
  * dhcp_native_ms
* Configuring DNS
  * dns_nsupdate
  * dns_nsupdate_gss
  * dns_libvirt
  * dns_dnscmd
  * dns_infoblox
  * dns_route53
  * dns_powerdns
* Configuring TFTP

The exact providers I mentioned are just examples. I'm not sure they all make sense in our documentation. For example, dns_dnscmd must run on Windows and we don't document Windows support. That's exactly why I think a reference manual in addition to our other manuals makes sense.

maximiliankolb commented on guides/common/assembly_performing-additional-configuration-on-capsule-server.adoc:
```diff
@@ -27,5 +27,9 @@ include::modules/proc_enabling-power-management-on-managed-hosts.adoc[leveloffse
 // Configuring DNS, DHCP, and TFTP on {SmartProxyServer}
 include::modules/proc_configuring-dns-dhcp-and-tftp.adoc[leveloffset=+1]
 
+include:modules/proc_configuring-dhcp-on-smart-proxies.adoc[leveloffset=+2]
+
+include:modules/proc_configuring-dns-on-smart-proxies.adoc[leveloffset=+2]
```
🤦 

Thanks, applied!

ekohl commented:
Since I'm not sure where else to write this, I'll give a quick high level overview.

The [Smart Proxy](https://github.com/theforeman/smart-proxy) is a small webservice that's really a framework. The real functionality is provided by modules. Each module implements what Foreman calls a Feature. If you navigate to the Smart Proxies detail page, you can see features such as `DNS`, `DHCP`, `Logs` and more. A bunch of these modules are shipped inside the `foreman-proxy` package (like `DNS`, `DHCP`), others are provided by plugins (like `Pulp`). The installer abstracts the packaging away for you, but this is important for a complete understanding.

Additionally, modules can have 0 or more providers. Providers are used when there are multiple implementations. For example, the DNS module is always the same: Foreman always uses the same API. Then each implementation is a provider. Providers can be explicitly configured (such as DNS) or programmatically (Puppet CA uses the `puppet_version` setting to determine the implementation). By convention we've named providers by their module: `$module_$provider`. This is only a convention and not hard enforced. Like modules, providers are shipped both in the `foreman-proxy` package and in plugin packages. `dns_nsupdate` is built in while `dns_infoblox` is provided by a separate package.

An example of a module without providers is TFTP. There is just a single implementation and it's inside the module.

As for configuration, modules and plugins can have configuration files. If so, they are named the same. So the `dns` module reads `dns.yml` while the `dns_nsupdate` provider reads `dns_nsupdate.yml`.

The installer abstracts this away from the user, which may make it harder to follow. However, we can let this return in the manual. This is why I suggested the structure in https://github.com/theforeman/foreman-documentation/pull/1440#pullrequestreview-1046221228. Does this help with a more fundamental understanding?

ekohl commented on guides/common/modules/proc_configuring-dhcp-native-ms.adoc:
```diff
@@ -0,0 +1,16 @@
+[id="Configuring_dhcp_native_ms_{context}"]
+= Configuring dhcp_native_ms
+
+The *dhcp_native_ms* plugin manages IP reservations in Microsoft Active Directory using its native API.
```
Perhaps it's not clear from this, but this must run on Windows. Given we only document Linux I wonder if we should document this.

In an ideal world we would have a guide "Configuring Smart Proxy on Windows" which would document how to run this and dns_dnscmd. Until then I'm leaning to dropping it.

ekohl commented on guides/common/modules/proc_configuring-dns-dnscmd.adoc:
```diff
@@ -0,0 +1,15 @@
+= Configuring dns_dnscmd
+
+While the `dns_nsupdate` provider creates dynamic records in Active Directory, the `dns_dnscmd` provider uses the `dnscmd` tool to create static DNS records in AD, which are not affected by scavenging.
+This requires that the {SmartProxy} is installed on a Windows server with `dnscmd` available.
```
Since we only document Linux setups today, should we also drop this?

ekohl commented on guides/common/modules/proc_configuring-dns-libvirt.adoc:
```diff
@@ -0,0 +1,19 @@
+It uses `ruby-libvirt` gem to connect to the local or remote instance of libvirt daemon.
+
+.Procedure
+. Edit `dns_libvirt.yml`:
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+# Libvirt network; only one network is supported.
+:network: default
+
+# Libvirt connection; ensure the 'foreman-proxy' user has permissions to connect.
+:url: qemu:///system
+----
```
Similar to `dhcp_libvirt`, we can use the installer:
```
foreman-installer \
  --foreman-proxy-dns true \
  --foreman-proxy-dns-provider libvirt \
  --foreman-proxy-libvirt-network default \
  --foreman-proxy-libvirt-url qemu:///system
```

And yes, you notice that you can only use a single network + URL if you use both dns_libvirt and dhcp_libvirt.

ekohl commented on guides/common/modules/proc_configuring-dns-nsupdate-gss.adoc:
```diff
@@ -0,0 +1,25 @@
+[id="Configuring_dns_nsupdate_gss_{context}"]
```
Doesn't this duplicate `guides/common/modules/proc_configuring-dynamic-dns-update-with-gss-tsig-authentication.adoc` (which itself needs a thorough readthrough)?

ekohl commented on guides/common/modules/proc_configuring-dns-nsupdate.adoc:
```diff
@@ -0,0 +1,25 @@
+[id="Configuring_dns_nsupdate_{context}"]
```
I'd say this is already covered in the existing guide.

ekohl commented on guides/common/modules/proc_configuring-isc-dhcp.adoc:
```diff
@@ -0,0 +1,51 @@
+[id="Configuring_isc_dhcp_{context}"]
```
This should be covered by the installer and the existing guides. I don't think we should document manual installation in the primary guide.

I've been wondering if we should document the reference bits on https://docs.theforeman.org (similar to how today it's on theforeman.org) or rather in [smart-proxy.git](https://github.com/theforeman/smart-proxy). The latter makes sense to me because it's more focused on developers. You can easily apply the same in a development setup.

maximiliankolb replied:
Removed as duplicate.

ekohl commented on guides/common/modules/proc_configuring-ms-dhcp.adoc:
```diff
@@ -0,0 +1,34 @@
+[id="Configuring_MS_DHCP_{context}"]
```
As noted before: do we want to document Windows? Right now I'm leaning to moving it over to smart-proxy.git.

maximiliankolb replied:
Moved for now; but as said before, I understand if we decide not to add this to the official Foreman documentation! (adding this is most certainly not my own idea...)

ekohl commented on guides/common/modules/proc_enabling-the-dhcp-module.adoc:
```diff
@@ -0,0 +1,28 @@
+[id="Enabling_the_DHCP_Module_{context}"]
```
I think this should be covered by our other guides already, by relying on the installer.

maximiliankolb replied:
Removed as duplicate.

ekohl commented on guides/common/modules/proc_enabling-the-dhcp-module.adoc:
```diff
@@ -0,0 +1,28 @@
+:server: 127.0.0.1
+----
++
+Note that if the DHCP server is running remotely, some providers (notably ISC) require that the configuration files must be accessible to the {SmartProxyServer}.
+You can use network file systems to achive this.
+. Optional: By default, you cna manage all available subnets.
+To mitigate the performance penalty from loading all available subnets, restrict DHCP to a list of subnets to optimize performace in `_My_Network_Address_/_My_Network_Mask_` notation:
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+:subnets: [192.168.205.0/255.255.255.128, 192.168.205.128/255.255.255.128]
+----
+. Optional: Adjust the configuration in the DHCP-provider specific configuration file.
```
Interestingly enough `guides/common/modules/proc_configuring-satellite-deployment-with-an-external-dhcp-server.adoc` doesn't document `--foreman-proxy-dhcp-subnets`. The equivalent of above would be:
```
--foreman-proxy-dhcp-subnets 192.168.205.0/255.255.255.128 \
--foreman-proxy-dhcp-subnets 192.168.205.128/255.255.255.128
```

I'm also very big on using RFC5735 IP ranges in documentation. See https://community.theforeman.org/t/using-ips-and-domains-in-tests-documentation/16600

maximiliankolb replied:
Added to "DHCP Options for Network Configuration" reference.

ekohl commented on guides/common/modules/proc_securing-the-dhcp-api.adoc:
```diff
@@ -0,0 +1,31 @@
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+# yum install -y bind
```
At least on EL8 you need bind-utils, not bind for `dnssec-keygen`. IIRC it's different on Debian.

maximiliankolb replied:
Fixed.

ekohl commented on guides/common/modules/proc_securing-the-dhcp-api.adoc:
```diff
@@ -0,0 +1,31 @@
+# dnssec-keygen -r /dev/urandom -a HMAC-MD5 -b 512 -n HOST omapi_key
+# cat Komapi_key.+*.private | grep ^Key|cut -d ' ' -f2-
+----
+. Edit your `/etc/dhcpd.conf`:
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+omapi-port 7911;
+key omapi_key {
+algorithm HMAC-MD5;
+  secret "XXXXXXXXX"; #<-The output from the generated key above.
+};
+omapi-key omapi_key;
+----
+. Ensure to add the omapi_key to `dhcp_isc.yml` on your {SmartProxyServer}.
+. Restart the `dhcpd` and `foreman-proxy` services.
+. In the {ProjectWebUI}, navigate to *Infrastructure > Smart Proxies*.
+. Select your {SmartProxy}.
+. On the *Overview* tab, click *Refresh* in the *Actions* menu.
```
The proper way to do this is using the installer:
```
foreman-installer \
  --foreman-proxy-dhcp-key-name "myname" \
  --foreman-proxy-dhcp-key-secret "mysecret"
```

We don't do this by default (because generating a secret and parsing that was hard) but surprisingly enough we don't document it either.

maximiliankolb replied:
Applied!

ekohl commented on guides/common/modules/proc_using-gss-tsig-dns.adoc:
```diff
@@ -0,0 +1,98 @@
+[id="Using_GSS_TSIG_DNS_{context}"]
```
This duplicates the `guides/common/modules/proc_configuring-dynamic-dns-update-with-gss-tsig-authentication.adoc` guide.

maximiliankolb replied:
Removed as duplicate.

ekohl reviewed (commented):
Thanks for working on this. I did a first pass on this.

maximiliankolb commented on guides/common/modules/proc_configuring-dns-on-smart-proxies.adoc:
```diff
@@ -0,0 +1,22 @@
+[id="Configuring_DNS_on_Smart_Proxies_{context}"]
```
Removed as duplicate.

ekohl commented:
The build looks red.

maximiliankolb commented:
Thank you so much @ekohl I really appreciate your review. I've tried to incorporate your feedback as much as possible.

maximiliankolb commented:
How can I proceed with my PR? Any other input? Is this good to merge?

ekohl commented:
I'll be honest and say that I haven't had time to read through the last update. Right now I'm working on getting the installer modules released so we can branch 3.4. However, it looks like I need to work on our CI nodes first because they have started to fail.

maximiliankolb commented:
Thanks for all the input so far! Do you have a rough timeline or would you like me to look for someone else to do a final review? content wise, not much/nothing has changed. I only restructured procedures to separate files.

maximiliankolb commented:
How can I proceed? Are there any objections to merging this? Can someone please review my PR?

ekohl commented on guides/common/modules/proc_securing-the-dhcp-api.adoc:
```diff
@@ -0,0 +1,26 @@
+--foreman-proxy-dhcp-key-name "_My_Name_" \
+--foreman-proxy-dhcp-key-secret "_My_Secret_"
+----
+. In the {ProjectWebUI}, navigate to *Infrastructure > Smart Proxies*.
+. Select your {SmartProxy}.
+. On the *Overview* tab, click *Refresh* in the *Actions* menu.
```
The installer will refresh if needed. In this particular case it's not even needed since it only affects dhcpd.
```suggestion
```

maximiliankolb replied:
Removed.

ekohl commented on guides/common/modules/ref_dhcp-options-for-network-configuration.adoc:
```diff
@@ -27,4 +27,8 @@ Set this to the address for {ProjectServer} on `eth1`.
   Sets the address of the DHCP server to manage.
 
+--foreman-proxy-dhcp-subnets::
+  Sets the subnets of the DHCP server to manage.
```
Should this mention that by default all subnets are managed? This helps to speed up things on large DHCP servers with many subnets, but the installer by default only supports a single subnet so I'm not sure this adds anything for most users. It'll only confuse them.

maximiliankolb replied:
I don't really know. Maybe someone else can weigh in?

Lennonka replied:
@ekohl What is the purpose of this argument in the first place? What is the scenario in which the user would prefer to use it?

ekohl replied:
Here is the example:
https://github.com/theforeman/smart-proxy/blob/9268440e7ea33d3fd0cae22896c403338e049cb9/config/settings.d/dhcp.yml.example#L11-L12
It mostly benefits the `dhcp_native_ms` provider (which the installer doesn't manage). I'm looking through the code, but can't find the code that actually ensures the whole setting is used.

I'd say that for most, if not all, users of the installer this parameter is not needed. Documenting it here will confuse users and make them do additional work. My suggestion is to leave it undocumented.

maximiliankolb replied:
Maybe I can just put it behind `ifdef::orcharhino[]`?

ekohl replied:
What is the benefit of including it in the Orcharhino manual? I'm quite sure it only affects `dhcp_native_ms` and that requires running on Windows. Has Orcharhino ported the installer to Windows?

I never looked at this before, but now as an installer maintainer I'm tempted to remove the option.

Lennonka replied:
Unless any product handles Windows, I'd go with undocumented.

maximiliankolb replied:
I have already removed anything regarding Windows for now. I'd be happy to push this upstream once a working installer exists.

ekohl replied:
Then please remove this part. I strongly insist on it.

maximiliankolb replied:
OK, changed from if not satellite to if orcharhino.

ekohl commented on guides/common/modules/con_managing-dhcp-using-smart-proxies.adoc:
```diff
@@ -0,0 +1,17 @@
+* `dhcp_infoblox` - Infoblox DHCP server
+* `dhcp_isc` - ISC DHCP server over OMAPI
+* `dhcp_remote_isc` - ISC DHCP server over OMAPI with leases mounted through networking
+* `dhcp_libvirt` - dnsmasq DHCP via libvirt API
```
```suggestion
ifndef::satellite[]
* `dhcp_libvirt` - dnsmasq DHCP via libvirt API
endif::[]
```

maximiliankolb replied:
Applied for libirt.

dhcp_native_ms is orcharhino only:
```
git grep -i -n -C 1 assembly_configuring-smart-proxy-on-microsoft-windows.adoc
guides/doc-Installing_Proxy/master.adoc-28-ifdef::orcharhino[]
guides/doc-Installing_Proxy/master.adoc:29:include::common/assembly_configuring-smart-proxy-on-microsoft-windows.adoc[leveloffset=+1]
guides/doc-Installing_Proxy/master.adoc-30-endif::[]
```

maximiliankolb commented on guides/common/assembly_managing-dhcp-and-dns-using-smart-proxies.adoc:
```diff
@@ -0,0 +1,11 @@
+include::modules/con_managing-dhcp-using-smart-proxies.adoc[]
+
+include::modules/proc_configuring-dhcp-libvirt.adoc[leveloffset=+1]
```
Applied.

maximiliankolb commented on guides/common/assembly_managing-dhcp-and-dns-using-smart-proxies.adoc:
```diff
@@ -0,0 +1,11 @@
+
+include::modules/proc_securing-the-dhcp-api.adoc[leveloffset=+1]
+
+include::modules/ref_dhcp-isc-settings.adoc[leveloffset=+1]
```
It's a reference file, but I've moved it to the appendix.

maximiliankolb commented on guides/common/modules/con_managing-dns-using-smart-proxies.adoc:
```diff
@@ -0,0 +1,19 @@
+You can use it to manage both Microsoft Active Directory and BIND servers.
+
+.Available DNS Providers
+* `dns_dnscmd` - static DNS records in Microsoft Active Directory
```
Made orcharhino-only.

maximiliankolb commented on guides/common/modules/con_managing-dns-using-smart-proxies.adoc:
```diff
@@ -0,0 +1,19 @@
+.Available DNS Providers
+* `dns_dnscmd` - static DNS records in Microsoft Active Directory
+* `dns_infoblox` - For more information, see {ProvisioningDocURL}Using_Infoblox_as_DHCP_and_DNS_Providers_provisioning[Using Infoblox as DHCP and DNS Providers] in _{ProvisioningDocTitle}_.
+* `dns_libvirt` - dnsmasq DNS via libvirt API
```
Applied!

maximiliankolb commented on guides/common/modules/proc_securing-the-dhcp-api.adoc:
```diff
@@ -0,0 +1,26 @@
+You can add a `omapi_key` to provide basic security.
+
+.Procedure
+. On Enterprise Linux 8, generate a key:
```
Applied.

maximiliankolb commented on guides/doc-Installing_Proxy/master.adoc:
```diff
@@ -23,6 +23,10 @@ include::common/assembly_performing-additional-configuration-on-capsule-server.a
 
+include::common/assembly_managing-dhcp-and-dns-using-smart-proxies.adoc[leveloffset=+1]
+
+include::common/assembly_configuring-smart-proxy-on-microsoft-windows.adoc[leveloffset=+1]
```
Made orcharhinon only.

maximiliankolb reviewed (commented):
Please have a look at my last commit.

ekohl commented on guides/common/modules/proc_securing-the-dhcp-api.adoc:
```diff
@@ -5,11 +5,23 @@ By default, the dhcpd API listens to any host.
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+ifndef::foreman-deb[]
+----
+# dnf install bind-utils
```
Would this work?
```suggestion
# {package-install} bind-utils
```

maximiliankolb replied:
Applied.

ekohl commented on guides/common/modules/proc_securing-the-dhcp-api.adoc:
```diff
@@ -0,0 +1,35 @@
+endif::[]
+ifdef::foreman-deb[]
+----
+# apt-get install bind9-utils
```
```suggestion
# {package-install} bind9-utils
```

maximiliankolb replied:
Applied.

maximiliankolb reviewed (commented):
Thanks again! Applied your suggestions.

ekohl commented on guides/common/modules/con_managing-dhcp-using-smart-proxies.adoc:
```diff
@@ -0,0 +1,16 @@
+{SmartProxies} can manage DHCP within their network.
+You can use the DHCP modules for {SmartProxy} to query for available IP addresses, add new, and delete existing reservations.
+Note that your {SmartProxy} cannot manage subnet declarations.
+ifdef::foreman-el,foreman-deb,katello[]
+For more information on managing subnet declarations, see https://github.com/theforeman/puppet-dhcp[puppet-dhcp].
+endif::[]
```
This ifdef block really belongs in a part of the manual, if any at all. Mostly because we don't have instructions about how to map the DHCP options in that module to what you can use in the installer. Right now I'm leaning to dropping it

ekohl replied:
This point is still open.

ekohl replied:
I think so yes.

maximiliankolb replied:
Removed.

ekohl reviewed (commented):
Structure wise this becomes chapter 5 and 6, but we also have chapter 3.3 which does the combined thing. Should become chapters 3.4(.z) and 3.5(.z) instead?

Structure wise there are some choices you can make.

First of all, go with feature as top level, as you did now. So
* DHCP
  * dhcp_libvirt
  * dhcp_isc
    * Securing the dhcpd api
  * dhcp_remote_isc
    * Set up external server
    * Configure Foreman server
  * ..
* DNS
  * dns_nsupdate
    * local DNS server
    * remote DNS server

The other is what's in the current guide, which focuses more on the scenario.

I like your new style, but it brings confusion to have both styles (as the PR currently does).

@Lennonka any thoughts on this?

maximiliankolb commented:
Rebased to master. I don't really have a strong opinion on the structure. If you have a good idea, please let me know. Alternatively, we can reorder content later on in a separate PR.

ekohl commented:
IMHO the structure is important if we cherry pick this, in particular to 3.1. This now has effects on the Satellite documentation

maximiliankolb commented:
@ekohl Your suggestion is to not create a new assembly, but rather include the content, where appropriate, in existing assemblies, isn't it?

ekohl commented on guides/common/assembly_managing-dhcp-and-dns-using-smart-proxies.adoc:
```diff
@@ -0,0 +1,11 @@
+
+include::modules/con_managing-dns-using-smart-proxies.adoc[]
+
+include::modules/proc_configuring-dns-libvirt.adoc[leveloffset=+1]
```
```suggestion
ifndef::satellite[]
include::modules/proc_configuring-dns-libvirt.adoc[leveloffset=+1]
endif::[]
```

maximiliankolb replied:
Applied.

ekohl commented:
I'd create an assembly per Smart Proxy feature with a provider. So you'd get the following assemblies:

* Preparing your Environment for Installation
* Installing `{SmartProxyServer}`
* Performing Additional Configuration on `{SmartProxyServer}` (but (possibly later) remove the DNS/DHCP part; TFTP should become its own chapter just like OpenSCAP and Power Management)
* Managing DHCP on `{SmartProxyServer}`
* Managing DNS on `{SmartProxyServer}`
* `{SmartProxyServer}` Scalability Considerations

So in short: if you see an installer argument `--foreman-proxy-$feature-provider` it should become its own assembly where each provider is a chapter(s). Simple features without providers belong under Additional Configuration.

Does that make sense? I know I'm suggesting to restructure a quite fundamental document so I'd like some buy in from actual writers.

maximiliankolb commented:
Restructured everything. Please see the last two commits.

ekohl commented on guides/common/modules/con_managing-dhcp-using-smart-proxies.adoc:
```diff
@@ -0,0 +1,21 @@
+
+The DHCP module can manage various available DHCP providers:
+
+* `dhcp_infoblox` - Infoblox DHCP server
```
I like what you did with with DNS and link to the "using ..." part. Could you do that here too?

maximiliankolb replied:
Applied!

ekohl commented on guides/common/modules/con_managing-dhcp-using-smart-proxies.adoc:
```diff
@@ -0,0 +1,21 @@
+
+* `dhcp_infoblox` - Infoblox DHCP server
+* `dhcp_isc` - ISC DHCP server over OMAPI
+* `dhcp_remote_isc` - ISC DHCP server over OMAPI with leases mounted through networking
```
This should link to the "Configuring ... with external DHCP" section since that documents it now.

maximiliankolb replied:
Applied.

ekohl commented on guides/common/modules/con_managing-dhcp-using-smart-proxies.adoc:
```diff
@@ -0,0 +1,21 @@
+The DHCP module can manage various available DHCP providers:
+
+* `dhcp_infoblox` - Infoblox DHCP server
+* `dhcp_isc` - ISC DHCP server over OMAPI
```
This should link to "Configuring DNS, DHCP, and TFTP on ..."

maximiliankolb replied:
Applied.

ekohl commented on guides/common/modules/con_managing-dns-using-smart-proxies.adoc:
```diff
@@ -0,0 +1,23 @@
+ifndef::satellite[]
+* `dns_libvirt` - dnsmasq DNS via libvirt API
+endif::[]
+* `dns_nsupdate_gss` - dynamic DNS update with GSS-TSIG
+* `dns_nsupdate` - dynamic DNS update using nsupdate
```
Can you swap these?
```suggestion
* `dns_nsupdate` - dynamic DNS update using nsupdate
* `dns_nsupdate_gss` - dynamic DNS update with GSS-TSIG
```

maximiliankolb replied:
Swapped.

ekohl commented on guides/common/modules/con_managing-dns-using-smart-proxies.adoc:
```diff
@@ -0,0 +1,23 @@
+Your {SmartProxy} can manage DNS records.
+DNS management contains updating and removing DNS records from existing DNS zones.
+
+After you have enabled DNS, your {SmartProxy} can manipulate any DNS server that complies with RFC 2136.
+You can use it to manage both Microsoft Active Directory and BIND servers.
```
This specifically talks about `dns_nsupdate` and `dns_nsupdate_gss`. Would it make sense to move it there?

maximiliankolb replied:
moved.

ekohl commented on guides/common/modules/con_managing-dns-using-smart-proxies.adoc:
```diff
@@ -0,0 +1,23 @@
+[id="Managing_DNS_Using_Smart_Proxies_{context}"]
+= Managing DNS Using {SmartProxies}
+
+Your {SmartProxy} can manage DNS records.
+DNS management contains updating and removing DNS records from existing DNS zones.
```
I was thinking about a generic bit of text. Something like:
```asciidoc
{Project} can manage DNS records using your {SmartProxy}.
DNS management contains updating and removing DNS records from existing DNS zones.
The {SmartProxy} has multiple providers, allowing you to integrate with your existing DNS infrastructure or deploy a new one.
```

A similar text can be used for the DHCP introduction.

maximiliankolb replied:
Applied and added.

ekohl reviewed (commented):
I noticed that the last link (currently an appendix) in the navigation is barely readable. Not limited to this guide, but this requires some CSS changes in our navigation.

On this PR: I like where it's going. The listing should be a bit more actionable, especially since some providers are documented somewhere else. See inline comments.

maximiliankolb reviewed (commented):
Rebased to master and applied all feedback possible in the last commit. (needs to be squashed before merging).

maximiliankolb commented:
Rebased to master and removed anything related to SmartProxy on Windows. Note that this branch needs to be squashed before merging.

Please re-review.

ekohl commented on guides/common/modules/con_managing-dns-using-smart-proxies.adoc:
```diff
@@ -0,0 +1,23 @@
+DNS management contains updating and removing DNS records from existing DNS zones.
+Your {SmartProxy} can use multiple providers, allowing you to integrate with your existing DNS infrastructure or deploy a new one.
+
+After you have enabled DNS, your {SmartProxy} can manipulate any DNS server that complies with RFC 2136.
```
This is only with the default provider.

background: [nsupdate](https://en.wikipedia.org/wiki/Nsupdate) is the tool used to implement the client side of [RFC 2136](https://datatracker.ietf.org/doc/html/rfc2136). That's why the provider is named `dns_nsupdate`.

Probably not the ideal text, but have you considered something like:
```suggestion
After you have enabled DNS, your {SmartProxy} can manipulate any DNS server that complies with RFC 2136 using the `dns_nsupdate` provider.
Other providers provide more direct integration, such as `dns_infoblox` for [Infoblox](https://www.infoblox.com/).
```

This would tie in to the last sentence of the previous paragraph.

maximiliankolb replied:
Thanks, applied!

ekohl commented on guides/common/modules/con_managing-dns-using-smart-proxies.adoc:
```diff
@@ -0,0 +1,23 @@
+* `dns_libvirt` - dnsmasq DNS via libvirt API
+endif::[]
+* `dns_nsupdate` - dynamic DNS update using nsupdate
+* `dns_nsupdate_gss` - dynamic DNS update with GSS-TSIG
```
This provider is documented in `guides/common/modules/proc_configuring-dynamic-dns-update-with-gss-tsig-authentication.adoc`. Would be good to link back.

maximiliankolb replied:
Added a link.

ekohl commented on guides/common/modules/con_managing-dhcp-using-smart-proxies.adoc:
```diff
@@ -0,0 +1,24 @@
+* `dhcp_infoblox` - For more information, see {ProvisioningDocURL}Using_Infoblox_as_DHCP_and_DNS_Providers_provisioning[Using Infoblox as DHCP and DNS Providers] in _{ProvisioningDocTitle}_.
+* `dhcp_isc` - ISC DHCP server over OMAPI.
+For more information, see {InstallingSmartProxyDocURL}configuring-dns-dhcp-and-tftp_smart-proxy[Configuring DNS, DHCP, and TFTP on {SmartProxyServer}] in _{InstallingSmartProxyDocTitle}_.
+* `dhcp_remote_isc` - ISC DHCP server over OMAPI with leases mounted through networking
```
```suggestion
* `dhcp_remote_isc` - ISC DHCP server over OMAPI with leases mounted through networking.
```

maximiliankolb replied:
Applied.

ekohl commented on guides/common/modules/proc_configuring-dns-libvirt.adoc:
```diff
@@ -0,0 +1,19 @@
+= Configuring dns_libvirt
+
+The _dns_libvirt_ DNS provider manages IP reservations and leases using dnsmasq through the libvirt API.
+It uses `ruby-libvirt` gem to connect to the local or remote instance of libvirt daemon.
```
```suggestion
It uses `ruby-libvirt` gem to connect to the local or a remote instance of libvirt daemon.
```

maximiliankolb replied:
Applied.

ekohl commented on guides/common/modules/ref_dhcp-options-for-network-configuration.adoc:
```diff
@@ -27,4 +27,8 @@ Set this to the address for {ProjectServer} on `eth1`.
 --foreman-proxy-dhcp-server::
```
I think this whole reference section needs redesign. It all only applies to the `dhcp_nsupdate` provider (except `--foreman-proxy-dhcp`), but the section doesn't mention that at all. Not a blocker for this PR, but this PR opens the path to make that sensible.

maximiliankolb replied:
Leaving this as is for now.

ekohl replied:
Yes, that's why I suggested to leave it out of the Satellite guide. Then it's safer for a cherry pick.

maximiliankolb replied:
I have now hidden it for build target satellite.

ekohl commented on guides/doc-Installing_Proxy/master.adoc:
```diff
@@ -23,7 +23,17 @@ include::common/assembly_performing-additional-configuration-on-capsule-server.a
 // {SmartProxyServer} Scalability Considerations
 include::common/modules/ref_capsule-server-scalability-considerations.adoc[leveloffset=+1]
+
+[appendix]
+include::common/modules/ref_dhcp-isc-settings.adoc[leveloffset=+1]
+
+[appendix]
+include::common/modules/ref_dhcp-options-for-network-configuration.adoc[leveloffset=+1]
```
I'm guessing this shouldn't be shown for satellite
```suggestion
ifndef::satellite[]
[appendix]
include::common/modules/ref_dhcp-isc-settings.adoc[leveloffset=+1]

[appendix]
include::common/modules/ref_dhcp-options-for-network-configuration.adoc[leveloffset=+1]
endif::[]
```

maximiliankolb replied:
Applied.

ekohl commented on guides/doc-Installing_Proxy/master.adoc:
```diff
@@ -23,7 +23,17 @@ include::common/assembly_performing-additional-configuration-on-capsule-server.a
 // Configuring {SmartProxyServer} with External Services
 include::common/assembly_configuring-external-services.adoc[leveloffset=+1]
 
+include::common/assembly_managing-dhcp-on-smart-proxies.adoc[leveloffset=+1]
+
+include::common/assembly_managing-dns-on-smart-proxies.adoc[leveloffset=+1]
```
Technically these options all also apply to the `doc-Installing_Server` side. I'd be OK with doing that in a follow up, but it's something to think about.

maximiliankolb replied:
👍

ekohl commented on guides/common/modules/con_managing-dns-using-smart-proxies.adoc:
```diff
@@ -0,0 +1,23 @@
+ifdef::orcharhino[]
+* `dns_dnscmd` - static DNS records in Microsoft Active Directory
+endif::[]
+* `dns_infoblox` - For more information, see {ProvisioningDocURL}Using_Infoblox_as_DHCP_and_DNS_Providers_provisioning[Using Infoblox as DHCP and DNS Providers] in _{ProvisioningDocTitle}_.
```
Naming thing: in the installer we don't use the `dns_` prefix, but rather automatically prefix it. To be concrete: the user passes in `--foreman-proxy-dns-provider nsupdate_gss` to use the `dns_nsupdate_gss` provider. What should we follow here?

The same applies to DHCP.

maximiliankolb replied:
For me, this is fine as is because there are no `foreman-installer` commands in this concept file. I have checked to ensure that all foreman-installer commands are without `[dhcp-|dns-]` prefix.

ekohl commented on guides/common/modules/con_managing-dhcp-using-smart-proxies.adoc:
```diff
@@ -0,0 +1,24 @@
+[id="Managing_DHCP_Using_Smart_Proxies_{context}"]
+= Managing DHCP Using {SmartProxies}
+
+{Project} can manage DHCP service using your {SmartProxy}.
```
I'd say `integrate with` is better. It can manage ISC DHCP but all others only talk using APIs.
```suggestion
{Project} can integrate with a DHCP service using your {SmartProxy}.
```

maximiliankolb replied:
Applied.

ekohl reviewed (commented):
When squashing, would it make sense to squash it in a DHCP and DNS commit?

maximiliankolb commented on guides/common/modules/con_managing-dhcp-using-smart-proxies.adoc:
```diff
@@ -0,0 +1,24 @@
+
+The DHCP module can manage various available DHCP providers:
+
+* `dhcp_infoblox` - For more information, see {ProvisioningDocURL}Using_Infoblox_as_DHCP_and_DNS_Providers_provisioning[Using Infoblox as DHCP and DNS Providers] in _{ProvisioningDocTitle}_.
```
Reading "long term", I'd say that this could be done in a follow-up PR.

ekohl replied:
Exactly.

maximiliankolb reviewed (commented):
Rebased to master, squashed into two commits, and applied all feedback where possible.

ekohl commented on guides/common/modules/ref_dhcp-isc-settings.adoc:
```diff
@@ -0,0 +1,43 @@
+
+The _dhcp_isc_ provider uses a combination of the ISC DHCP server OMAPI management interface and parsing of configuration and lease files.
+This requires it to be run either on the same host as the DHCP server or to have network filesystem access to these files.
+You can define the following settings in `dhcp_isc.yml`:
```
Rather than "you can define", should this be more generic?
```suggestion
The following settings are defined in `dhcp_isc.yml`:
```

My reasoning would be that the entire file is managed by the installer and in the manual the user is never supposed to touch these. That's also the reason why I'm very hesitant to include this in the first place. Documenting them gives the impression that it's supported to modify them, which I'd say is unsupported.

maximiliankolb replied:
Applied.

I was hoping to include this as a reference to let users know what could potentially be modified (using the installer).

ekohl replied:
I get that, and I do think it's a good idea to document at least somewhere. I'm questioning what the right place is though. There are a lot of examples config files (https://github.com/theforeman/smart-proxy/tree/develop/config/settings.d). Perhaps we should enhance those to have comments. The packaging simply takes these (and from plugins) and ships them by default. The installer then has its own templates, but we can base them on that.

Or we can look at creating a new reference guide altogether. I'd love to have a guide "What is the Smart Proxy" that starts with a high level overview for users/contributors who want to **understand** the concepts (rather than simply use it). We could also use those for onboarding new people.

Food for thought.

ekohl commented on guides/common/modules/ref_dhcp-isc-settings.adoc:
```diff
@@ -0,0 +1,43 @@
+You can define the following settings in `dhcp_isc.yml`:
+
+ifdef::foreman-el,katello[]
+.Configuring the path to the _config_ and _leases_ files on Enterprise Linux
```
```suggestion
.Configuring the path to the _config_ and _leases_ files on {EL}
```

Or just keep it generic (and apply the same below to Debian):
```suggestion
.Configuring the path to the _config_ and _leases_ files
```

maximiliankolb replied:
Made it more generic.

ekohl commented on guides/common/modules/ref_dhcp-isc-settings.adoc:
```diff
@@ -0,0 +1,43 @@
+= dhcp_isc Settings
+
+The _dhcp_isc_ provider uses a combination of the ISC DHCP server OMAPI management interface and parsing of configuration and lease files.
+This requires it to be run either on the same host as the DHCP server or to have network filesystem access to these files.
```
We have `dhcp_remote_isc` for the network access.
```suggestion
This requires it to be run either on the same host as the DHCP server.
```

maximiliankolb replied:
Thanks, applied!

ekohl commented on guides/common/modules/ref_dhcp-isc-settings.adoc:
```diff
@@ -0,0 +1,43 @@
+.Setting the host on which the DHCP server runs on
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+:server: _My_DHCP_Server_FQDN_
```
`server` is in `dhcp.yml`, not `dhcp_isc.yml`.

maximiliankolb replied:
Thanks, applied!

maximiliankolb commented on guides/common/modules/proc_configuring-dns-libvirt.adoc:
```diff
@@ -0,0 +1,19 @@
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+foreman-installer \
```
Fixed.

maximiliankolb reviewed (commented):
Applied more feedback and rebased to master. I am unsure about one thread which I have kept open.

ekohl commented on guides/common/modules/con_managing-dhcp-using-smart-proxies.adoc:
```diff
@@ -0,0 +1,21 @@
+[id="Managing_DHCP_Using_Smart_Proxies_{context}"]
+= Managing DHCP Using {SmartProxies}
+
+{Project} can integrate with a DHCP service using your {SmartProxy}.
```
Perhaps a writer could review this, but the title is plural (SmartProxies) and this sentence is singular. Should it be consistent? `your {singular}` implies there is one while you can use any number of them.

maximiliankolb replied:
Changed to "SmartProxies" attribute.

maximiliankolb replied:
Please close this thread if you agree @Lennonka

ekohl commented on guides/common/modules/proc_securing-the-dhcp-api.adoc:
```diff
@@ -0,0 +1,35 @@
+= Securing the dhcpd API
+
+By default, the dhcpd API listens to any host.
+You can add a `omapi_key` to provide basic security.
```
```suggestion
You can add an `omapi_key` to provide basic security.
```

maximiliankolb replied:
Applied.

ekohl commented on guides/common/modules/con_managing-dhcp-using-smart-proxies.adoc:
```diff
@@ -0,0 +1,21 @@
+= Managing DHCP Using {SmartProxies}
+
+{Project} can integrate with a DHCP service using your {SmartProxy}.
+The {SmartProxy} has multiple providers, allowing you to integrate with your existing DHCP infrastructure or deploy a new one.
```
Should be generic?
```suggestion
A {SmartProxy} has multiple providers, allowing you to integrate with your existing DHCP infrastructure or deploy a new one.
```

Also, would it make sense to mention that one `{SmartProxy}` can only use one provider active at a time?

maximiliankolb replied:
Applied. IMHO only one service is a given due to line 4.

ekohl reviewed (commented):
At this point I'm happy with it on both the technical and structural level. Some minor textual comments, but other than that I'd ACK this.

I'd really appreciate a professional writer to review the exact contents. @Lennonka could you take a look at the generated texts?

maximiliankolb reviewed (commented):
Applied your feedback.

Thank you so much @ekohl for your endless patience and truely helpful comments!

Lennonka commented:
For some reason, the Installing Proxy guide doesn't build correctly for orcharhino on my machine. @maximiliankolb Is it working well on yours?
My situation is this: The HTML build contains only the first module of the first assembly, everything else is missing in the output.
I.e. I can only see
1. Preparing your Environment for Installation
1.1. System Requirements
Maybe an unclosed ifdef statement?

Lennonka commented on guides/common/modules/con_managing-dhcp-using-smart-proxies.adoc:
```diff
@@ -0,0 +1,21 @@
+[id="Managing_DHCP_Using_Smart_Proxies_{context}"]
+= Managing DHCP Using {SmartProxies}
+
+{Project} can integrate with a DHCP service using {SmartProxies}.
```
Sorry to say this, but I'd prefer it the other way around - the singular incl. in the title, ID and filename.

ekohl replied:
It was the inconsistency that bothered me. Singular is fine too.

maximiliankolb replied:
Changed title, anchor, and file name to singular for both dhcp and dns.

maximiliankolb commented:
> For some reason, the Installing Proxy guide doesn't build correctly for orcharhino on my machine. @maximiliankolb Is it working well on yours? My situation is this: The HTML build contains only the first module of the first assembly, everything else is missing in the output. I.e. I can only see
> 
> 1. Preparing your Environment for Installation
>    1.1. System Requirements
>    My output ends with the list item `* qdrouterd`.
> 
> Maybe something wrong with a conditional statement?

Yes, it's broken for some reason. I will have a look at this in some other PR. Thanks for letting me know. In downstream, I don't use any installing guide at the moment, so it's not something I actively monitor.

Rebased to master. But let me know if you'd like to see additional changes or if this is good to be merged. :v:

maximiliankolb commented:
@Lennonka Would you mind having another look at my PR?

goarsna commented on guides/common/modules/ref_dhcp-isc-settings.adoc:
```diff
@@ -0,0 +1,45 @@
+= dhcp_isc Settings
+
+The _dhcp_isc_ provider uses a combination of the ISC DHCP server OMAPI management interface and parsing of configuration and lease files.
+This requires it to be run either on the same host as the DHCP server.
```
There is an **or** missing for the **either**.

maximiliankolb replied:
Reworded.

goarsna commented on guides/common/modules/ref_dhcp-isc-settings.adoc:
```diff
@@ -0,0 +1,45 @@
+This requires it to be run either on the same host as the DHCP server.
+The following settings are defined in `dhcp_isc.yml`:
+
+ifdef::foreman-el,katello[]
```
`orcharhino` is missing here.

```suggestion
ifdef::foreman-el,katello,orcharhino[]
```

maximiliankolb replied:
Fixed.

goarsna reviewed (changes requested):
I had a look at the changes and made two suggestions. Despite that everything LGTM.

maximiliankolb commented:
Applied all feedback by @goarsna and rebased to master.

Still waiting for a final approval.

ekohl reviewed (approved):
I'm happy with this, but I'd ask you to submit a PR to 3.1 so we can easily see the diff there. This is mostly because there may be support implications for Satellite on a released product. In case you weren't aware, for Satellite having it documented means it's supported.

@Lennonka do you want to have a last read through before we merge it?

Lennonka commented on guides/common/modules/con_managing-dhcp-using-smart-proxy.adoc:
```diff
@@ -0,0 +1,21 @@
+[id="Managing_DHCP_Using_Smart_Proxy_{context}"]
+= Managing DHCP Using {SmartProxy}
+
+{Project} can integrate with a DHCP service using {SmartProxies}.
```
Singular

Lennonka commented on guides/common/modules/con_managing-dhcp-using-smart-proxy.adoc:
```diff
@@ -0,0 +1,21 @@
+= Managing DHCP Using {SmartProxy}
+
+{Project} can integrate with a DHCP service using {SmartProxies}.
+A {SmartProxy} has multiple providers, allowing you to integrate with your existing DHCP infrastructure or deploy a new one.
```
`A {SmartProxy} has multiple DHCP service providers that you can use to integrate {Project} with your existing DHCP infrastructure or deploy a new one.` maybe?

Lennonka commented on guides/common/modules/con_managing-dhcp-using-smart-proxy.adoc:
```diff
@@ -0,0 +1,21 @@
+You can use the DHCP module of {SmartProxy} to query for available IP addresses, add new, and delete existing reservations.
+Note that your {SmartProxy} cannot manage subnet declarations.
+
+The DHCP module can manage various available DHCP providers:
```
"implements" instead of "can manage"?

Lennonka replied:
Or maybe just a phony heading like `.Available DHCP Providers` instead of this sentence? (same as for DNS below)

Lennonka commented on guides/common/modules/con_managing-dns-using-smart-proxy.adoc:
```diff
@@ -0,0 +1,27 @@
+
+{Project} can manage DNS records using your {SmartProxy}.
+DNS management contains updating and removing DNS records from existing DNS zones.
+Your {SmartProxy} can use multiple providers, allowing you to integrate with your existing DNS infrastructure or deploy a new one.
```
Perhaps a tiny rephrase as above? :)

Lennonka commented on guides/common/modules/proc_securing-the-dhcp-api.adoc:
```diff
@@ -0,0 +1,35 @@
+[id="Securing_the_dhcpd_API_{context}"]
+= Securing the dhcpd API
+
```
WTF is dhcpd API, how does it relate to a DHCP provider or SmartProxy and why should I care about securing it? :D

ekohl replied:
I think this is a bit below:
> By default, the dhcpd API listens to any host.
> You can add an `omapi_key` to provide basic security.

Should this be phrased stronger? Like

```
By default, the dhcpd API listens to any host without access controls.
```

maximiliankolb replied:
I have applied your suggestion.

Lennonka reviewed (changes requested):
Oki, I tried to give it another review. However, when the output doesn't generate properly, that's it.

Lennonka reviewed (approved):
Good to go!

maximiliankolb commented:
Thanks for reviewing Ewound and Lena! Merged to master and cherry-picked: ✅ 
   a7ab3465d..777670ae7  3.4 -> 3.4
   f2d48fbcb..22b1d9ff6  3.3 -> 3.3
   62d1789ec..0778a7b15  3.2 -> 3.2
   b467691ee..1c434d211  3.1 -> 3.1

ekohl commented:
I had hoped that for 3.1 we'd have a PR so we could see the changes, but I'm glad we got this in. Thanks for the big effort @maximiliankolb! This is IMHO the base to build on.

ekohl commented:
> This is IMHO the base to build on.

I started with #1607.
